Add support for using 'currentcolor' with color-mix()

weinig opened the pull request:
#### f99a53e73e5181fada3cf04e56adfa01a2e05563
<pre>
Add support for using &apos;currentcolor&apos; with color-mix()
<a href="https://bugs.webkit.org/show_bug.cgi?id=250459">https://bugs.webkit.org/show_bug.cgi?id=250459</a>

Reviewed by Darin Adler.

To support &apos;currentcolor&apos; as the origin color in a color-mix()
value, this change adds a new CSSPrimitiveValue type called
CSSUnresolvedColor as a peer to the existing color value, that
represents, as the name suggests, unresolved color values.

In this change, CSSUnresolvedColor is only used for color-mix(),
but it is implemented in a way so that we can add support for
other unresolved colors like relative colors.

In addition to CSSUnresolvedColor, to fully support &apos;currentcolor&apos;
StyleColor was updated to add a third kind in addition to Color
and CurrentColor, for color-mix(). This allows for us to do late
evaluation of the color-mix() in the case of a &apos;currentcolor&apos;
(or even in more interesting cases like an origin which is another
color-mix that itself has &apos;currentcolor&apos; as the origin).

CSSUnresolvedColor also fixes an issue where we were not preserving
calc() values for the percentages in color-mix() in the specified
value serialization.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
Add new files.

* Source/WebCore/css/CSSPrimitiveValue.cpp:
(WebCore::isValidCSSUnitTypeForDoubleConversion):
(WebCore::isStringType):
(WebCore::CSSPrimitiveValue::init):
(WebCore::CSSPrimitiveValue::cleanup):
(WebCore::CSSPrimitiveValue::unitTypeString):
(WebCore::CSSPrimitiveValue::formatNumberForCustomCSSText const):
(WebCore::CSSPrimitiveValue::equals const):
* Source/WebCore/css/CSSPrimitiveValue.h:
(WebCore::isValueID):
* Source/WebCore/css/CSSUnits.cpp:
(WebCore::unitCategory):
(WebCore::operator&lt;&lt;):
* Source/WebCore/css/CSSUnits.h:
* Source/WebCore/css/calc/CSSCalcCategoryMapping.cpp:
(WebCore::hasDoubleValue):
Add new CSS_UNRESOLVED_COLOR kind.

* Source/WebCore/css/StyleColor.cpp:
(WebCore::StyleColor::copy):
(WebCore::StyleColor::resolveColor const):
(WebCore::StyleColor::isCurrentColor const):
(WebCore::StyleColor::isColorMix const):
(WebCore::StyleColor::resolveAbsoluteComponents):
(WebCore::resolveAbsoluteComponents):
(WebCore::resolveColor):
(WebCore::serializationForCSS):
(WebCore::operator&lt;&lt;):
* Source/WebCore/css/StyleColor.h:
(WebCore::StyleColor::StyleColor):
(WebCore::StyleColor::operator=):
(WebCore::StyleColor::currentColor):
(WebCore::operator==):
(WebCore::operator!=):
(WebCore::StyleColor::operator== const):
(WebCore::StyleColor::operator!= const):
(WebCore::CurrentColor::operator== const): Deleted.
(WebCore::CurrentColor::operator!= const): Deleted.
Support a new StyleColorMix kind. Rename CurrentColor to StyleCurrentColor
for consistency.

* Source/WebCore/css/color/CSSResolvedColorMix.cpp: Added.
(WebCore::normalizedMixPercentages):
(WebCore::mixColorComponentsUsingColorInterpolationMethod):
(WebCore::mix):
* Source/WebCore/css/color/CSSResolvedColorMix.h: Added.
CSSResolvedColorMix serves as the implementation of the final
resolved color-mix(), and is used for eager mixing (in the case
of the Raw variant currently used on workers) and in StyleColor
when doing the final resolution.

* Source/WebCore/css/color/CSSUnresolvedColor.cpp: Added.
(WebCore::CSSUnresolvedColor::cssText const):
(WebCore::CSSUnresolvedColor::equals const):
(WebCore::CSSUnresolvedColor::createStyleColor const):
* Source/WebCore/css/color/CSSUnresolvedColor.h: Added.
CSSUnresolvedColor serves a container for unresolved colors
in CSSPrimitiveValue. For now it only supports color-mix(),
but is setup to support other kinds going future.

* Source/WebCore/css/color/CSSUnresolvedColorMix.cpp: Added.
(WebCore::serializationForCSS):
(WebCore::operator==):
(WebCore::operator!=):
(WebCore::createStyleColor):
* Source/WebCore/css/color/CSSUnresolvedColorMix.h: Added.
Value type for unresolved color-mix(), used as the &quot;specified
value&quot; for color-mix() to correctly serialize per-spec and
correctly resolve to a structured StyleColor at style building
time.

* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeOriginColorRaw):
(WebCore::CSSPropertyParserHelpers::consumeOptionalAlphaRaw):
(WebCore::CSSPropertyParserHelpers::consumeOptionalAlphaRawAllowingSymbolTableIdent):
(WebCore::CSSPropertyParserHelpers::parseRelativeRGBParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseNonRelativeRGBParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseRGBParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseRelativeHSLParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseNonRelativeHSLParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseHSLParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseHWBParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseRelativeHWBParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseNonRelativeHWBParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseLabParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseRelativeLabParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseNonRelativeLabParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseLCHParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseRelativeLCHParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseNonRelativeLCHParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseColorFunctionForRGBTypesRaw):
(WebCore::CSSPropertyParserHelpers::parseRelativeColorFunctionForRGBTypes):
(WebCore::CSSPropertyParserHelpers::parseColorFunctionForXYZTypesRaw):
(WebCore::CSSPropertyParserHelpers::parseRelativeColorFunctionForXYZTypes):
(WebCore::CSSPropertyParserHelpers::parseRelativeColorFunctionParameters):
(WebCore::CSSPropertyParserHelpers::parseNonRelativeColorFunctionParameters):
(WebCore::CSSPropertyParserHelpers::parseColorFunctionParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseColorContrastFunctionParametersRaw):
(WebCore::CSSPropertyParserHelpers::consumeColorMixComponentRaw):
(WebCore::CSSPropertyParserHelpers::consumeColorMixComponent):
(WebCore::CSSPropertyParserHelpers::parseColorMixFunctionParametersRaw):
(WebCore::CSSPropertyParserHelpers::hasNonCalculatedZeroPercentage):
(WebCore::CSSPropertyParserHelpers::parseColorMixFunctionParameters):
(WebCore::CSSPropertyParserHelpers::parseColorFunctionRaw):
(WebCore::CSSPropertyParserHelpers::parseColorFunction):
(WebCore::CSSPropertyParserHelpers::consumeColorWorkerSafe):
(WebCore::CSSPropertyParserHelpers::consumeColor):
(WebCore::CSSPropertyParserHelpers::consumeOriginColor): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeOptionalAlpha): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeOptionalAlphaAllowingSymbolTableIdent): Deleted.
(WebCore::CSSPropertyParserHelpers::parseRelativeRGBParameters): Deleted.
(WebCore::CSSPropertyParserHelpers::parseNonRelativeRGBParameters): Deleted.
(WebCore::CSSPropertyParserHelpers::parseRGBParameters): Deleted.
(WebCore::CSSPropertyParserHelpers::parseRelativeHSLParameters): Deleted.
(WebCore::CSSPropertyParserHelpers::parseNonRelativeHSLParameters): Deleted.
(WebCore::CSSPropertyParserHelpers::parseHSLParameters): Deleted.
(WebCore::CSSPropertyParserHelpers::parseHWBParameters): Deleted.
(WebCore::CSSPropertyParserHelpers::parseRelativeHWBParameters): Deleted.
(WebCore::CSSPropertyParserHelpers::parseNonRelativeHWBParameters): Deleted.
(WebCore::CSSPropertyParserHelpers::parseLabParameters): Deleted.
(WebCore::CSSPropertyParserHelpers::parseRelativeLabParameters): Deleted.
(WebCore::CSSPropertyParserHelpers::parseNonRelativeLabParameters): Deleted.
(WebCore::CSSPropertyParserHelpers::parseLCHParameters): Deleted.
(WebCore::CSSPropertyParserHelpers::parseRelativeLCHParameters): Deleted.
(WebCore::CSSPropertyParserHelpers::parseNonRelativeLCHParameters): Deleted.
(WebCore::CSSPropertyParserHelpers::parseColorFunctionForRGBTypes): Deleted.
(WebCore::CSSPropertyParserHelpers::parseColorFunctionForXYZTypes): Deleted.
(WebCore::CSSPropertyParserHelpers::parseColorFunctionParameters): Deleted.
(WebCore::CSSPropertyParserHelpers::parseColorContrastFunctionParameters): Deleted.
(WebCore::CSSPropertyParserHelpers::normalizedMixPercentages): Deleted.
(WebCore::CSSPropertyParserHelpers::mixColorComponentsUsingColorInterpolationMethod): Deleted.
(WebCore::CSSPropertyParserHelpers::mixColorComponents): Deleted.
Renames color parsing functions to use the Raw suffix to match convention and
make it clear that the new color-mix parsing function is sufficiently different
in that it is not eagerly resolving.

* Source/WebCore/platform/graphics/ColorInterpolationMethod.cpp:
(WebCore::serializationForCSS):
* Source/WebCore/platform/graphics/ColorInterpolationMethod.h:
Add CSS serialization support and fix HasHueInterpolationMethod to actually
match the right structs.

* Source/WebCore/style/ColorFromPrimitiveValue.cpp:
(WebCore::Style::colorFromPrimitiveValue):
Support CSSUnresolvedColor CSSPrimitiveValues when building StyleColors.

Canonical link: <a href="https://commits.webkit.org/258970@main">https://commits.webkit.org/258970@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/68841301a1d64c0c37852b213b3b5453e76ce424

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103506 "16 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12624 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/36466 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/112742 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/172951 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/107460 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13654 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/3526 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/95761 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/111920 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/109279 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/10508 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/93585 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/38245 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/92329 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25172 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/79889 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6009 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/26575 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6187 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/3102 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12169 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46084 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6162 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7941 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/CSSPrimitiveValue.cpp:
```diff
@@ -489,6 +492,13 @@ void CSSPrimitiveValue::init(RefPtr<CSSCalcValue>&& c)
     m_value.calc = c.leakRef();
 }
 
+void CSSPrimitiveValue::init(Ref<CSSUnresolvedColor>&& unresolvedColor)
```
We need to start using constructors instead of init functions for this sort of thing. You will see that in my next patch which will conflict with this one a bit. I think you could just go ahead and do this for this new type even though the others will be in my change.

See https://github.com/darinadler/WebKit/commit/b2b9e1826f5886ebbbe361e9c4bcf8af478f687f for my work in progress.

darinadler commented on Source/WebCore/css/CSSPrimitiveValue.cpp:
```diff
@@ -1425,6 +1439,8 @@ ALWAYS_INLINE String CSSPrimitiveValue::formatNumberForCustomCSSText() const
     case CSSUnitType::CSS_URI:
         return serializeURL(m_value.string);
+    case CSSUnitType::CSS_UNRESOLVED_COLOR:
+        return m_value.unresolvedColor->cssText();
```
Do these functions really need to be named `cssText`? Just because of a peculiar name choice in the web exposed function.

weinig replied:
Nope. How about serializationForCSS to match the other color serialization functions.

darinadler commented on Source/WebCore/css/StyleColor.h:
```diff
@@ -110,25 +115,107 @@ class StyleColor {
 };
 
+struct StyleColorMix {
+    WTF_MAKE_FAST_ALLOCATED;
```
There's a flavor of this macro for structs that you can use so you don't have to explicitly say `public:` after it.

weinig replied:
Oh nice. I didn't think to look. Thanks.

darinadler commented on Source/WebCore/css/StyleColor.h:
```diff
@@ -110,25 +115,107 @@ class StyleColor {
+    return !(a == b);
+}
+
+inline bool operator==(const StyleCurrentColor&, const StyleCurrentColor&)
```
`constexpr`

darinadler commented on Source/WebCore/css/StyleColor.h:
```diff
@@ -110,25 +115,107 @@ class StyleColor {
+    Component mixComponents2;
+};
+
+inline bool operator==(const StyleColorMix::Component& a, const StyleColorMix::Component& b)
```
`constexpr`

weinig replied:
I don't think this one can be, since it would require operator== for Color to be constexpr, which it is not at the moment.

darinadler commented on Source/WebCore/css/StyleColor.h:
```diff
@@ -110,25 +115,107 @@ class StyleColor {
+        && a.percentage == b.percentage;
+}
+
+inline bool operator!=(const StyleColorMix::Component& a, const StyleColorMix::Component& b)
```
`constexpr`

darinadler commented on Source/WebCore/css/StyleColor.h:
```diff
@@ -110,25 +115,107 @@ class StyleColor {
+    return !(a == b);
+}
+
+inline bool operator==(const StyleColorMix& a, const StyleColorMix& b)
```
`constexpr`

darinadler commented on Source/WebCore/css/StyleColor.h:
```diff
@@ -110,25 +115,107 @@ class StyleColor {
+        && a.mixComponents2 == b.mixComponents2;
+}
+
+inline bool operator!=(const StyleColorMix& a, const StyleColorMix& b)
```
`constexpr`

darinadler commented on Source/WebCore/css/StyleColor.h:
```diff
@@ -110,25 +115,107 @@ class StyleColor {
+    return true;
+}
+
+inline bool operator!=(const StyleCurrentColor&, const StyleCurrentColor&)
```
`constexpr`

darinadler commented on Source/WebCore/css/StyleColor.h:
```diff
@@ -110,25 +115,107 @@ class StyleColor {
+    return false;
+}
+
+inline bool StyleColor::operator==(const StyleColor& other) const
```
`constexpr`

Also it seems like this could be non-member.

darinadler commented on Source/WebCore/css/StyleColor.h:
```diff
@@ -110,25 +115,107 @@ class StyleColor {
+    return m_color == other.m_color;
+}
+
+inline bool StyleColor::operator!=(const StyleColor& other) const
```
`constexpr`

Also it seems like this could be non-member.

darinadler commented on Source/WebCore/css/color/CSSUnresolvedColor.cpp:
```diff
@@ -0,0 +1,54 @@
+
+String CSSUnresolvedColor::cssText() const
+{
+    return WTF::switchOn(m_value, [] (const auto& unresolved) -> String { return WebCore::serializationForCSS(unresolved); });
```
Do you need the `const`?

darinadler commented on Source/WebCore/css/color/CSSUnresolvedColorMix.cpp:
```diff
@@ -0,0 +1,99 @@
+        builder.append(' ', component.percentage->customCSSText());
+}
+
+String serializationForCSS(const CSSUnresolvedColorMix& unresolved)
```
Maybe this should be a pair of functions. Always have the one that takes the `SteingBuilder` and the one without the `StringBuilder` that returns the `String`.

weinig replied:
That seems like the right pattern. Would be interesting to see if we could make one generate efficiently from the other (or both from a more generic third) at some point as well.

darinadler commented on Source/WebCore/platform/graphics/ColorInterpolationMethod.cpp:
```diff
@@ -26,10 +26,78 @@
 
 namespace WebCore {
 
+void serializationForCSS(StringBuilder& builder, ColorInterpolationColorSpace interpolationColorSpace)
```
Should have a function that returns an `ASCIILiteral` separate from the one that does the appending.

mdubet commented on Source/WebCore/css/CSSPrimitiveValue.h:
```diff
@@ -411,6 +417,16 @@ inline bool isValueID(const CSSPrimitiveValue* value, CSSValueID id)
     return valueID(value) == id;
 }
 
+inline bool isValueID(const RefPtr<CSSPrimitiveValue>& value, CSSValueID id)
```
It's not the role of `CSSPrimitiveValue` to know about all the different kinds of pointers/refs in the codebase. This should be handled at caller site maybe?

weinig replied:
I am not sure I agree with that sentiment. Overloading here makes the call sites much easier to read.

mdubet commented on Source/WebCore/css/CSSPrimitiveValue.cpp:
```diff
@@ -135,6 +136,7 @@ static inline bool isValidCSSUnitTypeForDoubleConversion(CSSUnitType unitType)
     case CSSUnitType::CSS_STRING:
     case CSSUnitType::CSS_UNICODE_RANGE:
     case CSSUnitType::CSS_UNKNOWN:
+    case CSSUnitType::CSS_UNRESOLVED_COLOR:
```
This "unresolved color" concept seems a bit odd in the CSS Value part of the codebase. `color-mix`, `currentcolor`, ... are all actual CSS values that we can find in a CSS file, but not "unresolved color". It seems that this is a rendering concept which should only exist later in the pipeline (in RenderStyle).

weinig replied:
The unresolved color concept is needed to support correct serialization of calc() in the specified value serialization. I would be happy to debate the name, but this seemed to have the cleanest split for me.

mdubet commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -3155,6 +3169,62 @@ static Color parseColorFunction(CSSParserTokenRange& range, const CSSParserConte
     return color;
 }
 
+static std::optional<ColorOrUnresolvedColor> parseColorFunction(CSSParserTokenRange& range, const CSSParserContext& context)
```
The duplication of the big switch statement for each color function seems suspicious.

weinig replied:
The worker specific path is an odd duck. Keeping that working requires a bit of duplication at the root.

mdubet commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -3189,15 +3259,22 @@ RefPtr<CSSPrimitiveValue> consumeColor(CSSParserTokenRange& range, const CSSPars
-            return nullptr;
+        return CSSValuePool::singleton().createColorValue(Color { *parsedColor });
+
+    if (auto colorOrUnresolvedColor = parseColorFunction(range, context)) {
```
👍 Nice way to handle the resolved/unresolved parsing result

mdubet commented on Source/WebCore/css/color/CSSResolvedColorMix.h:
```diff
@@ -0,0 +1,47 @@
+
+namespace WebCore {
+
+struct CSSResolvedColorMix {
```
Having both resolved/unresolved for each kind of color aggregate will be very heavy to maintain (color-mix, color-contrast, relative colors, ...)

weinig replied:
I don't think there is really a way around it as they need to store different types and have different serialization rules.

mdubet reviewed (commented):
I've tried a similar patch with a slightly different approach here : https://github.com/WebKit/WebKit/pull/8701

Could you take a look before landing this one ?

Commit:
Add support for using 'currentcolor' with color-mix()
https://bugs.webkit.org/show_bug.cgi?id=250459

Reviewed by Darin Adler.

To support 'currentcolor' as the origin color in a color-mix()
value, this change adds a new CSSPrimitiveValue type called
CSSUnresolvedColor as a peer to the existing color value, that
represents, as the name suggests, unresolved color values.

In this change, CSSUnresolvedColor is only used for color-mix(),
but it is implemented in a way so that we can add support for
other unresolved colors like relative colors.

In addition to CSSUnresolvedColor, to fully support 'currentcolor'
StyleColor was updated to add a third kind in addition to Color
and CurrentColor, for color-mix(). This allows for us to do late
evaluation of the color-mix() in the case of a 'currentcolor'
(or even in more interesting cases like an origin which is another
color-mix that itself has 'currentcolor' as the origin).

CSSUnresolvedColor also fixes an issue where we were not preserving
calc() values for the percentages in color-mix() in the specified
value serialization.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
Add new files.

* Source/WebCore/css/CSSPrimitiveValue.cpp:
(WebCore::isValidCSSUnitTypeForDoubleConversion):
(WebCore::isStringType):
(WebCore::CSSPrimitiveValue::init):
(WebCore::CSSPrimitiveValue::cleanup):
(WebCore::CSSPrimitiveValue::unitTypeString):
(WebCore::CSSPrimitiveValue::formatNumberForCustomCSSText const):
(WebCore::CSSPrimitiveValue::equals const):
* Source/WebCore/css/CSSPrimitiveValue.h:
(WebCore::isValueID):
* Source/WebCore/css/CSSUnits.cpp:
(WebCore::unitCategory):
(WebCore::operator<<):
* Source/WebCore/css/CSSUnits.h:
* Source/WebCore/css/calc/CSSCalcCategoryMapping.cpp:
(WebCore::hasDoubleValue):
Add new CSS_UNRESOLVED_COLOR kind.

* Source/WebCore/css/StyleColor.cpp:
(WebCore::StyleColor::copy):
(WebCore::StyleColor::resolveColor const):
(WebCore::StyleColor::isCurrentColor const):
(WebCore::StyleColor::isColorMix const):
(WebCore::StyleColor::resolveAbsoluteComponents):
(WebCore::resolveAbsoluteComponents):
(WebCore::resolveColor):
(WebCore::serializationForCSS):
(WebCore::operator<<):
* Source/WebCore/css/StyleColor.h:
(WebCore::StyleColor::StyleColor):
(WebCore::StyleColor::operator=):
(WebCore::StyleColor::currentColor):
(WebCore::operator==):
(WebCore::operator!=):
(WebCore::StyleColor::operator== const):
(WebCore::StyleColor::operator!= const):
(WebCore::CurrentColor::operator== const): Deleted.
(WebCore::CurrentColor::operator!= const): Deleted.
Support a new StyleColorMix kind. Rename CurrentColor to StyleCurrentColor
for consistency.

* Source/WebCore/css/color/CSSResolvedColorMix.cpp: Added.
(WebCore::normalizedMixPercentages):
(WebCore::mixColorComponentsUsingColorInterpolationMethod):
(WebCore::mix):
* Source/WebCore/css/color/CSSResolvedColorMix.h: Added.
CSSResolvedColorMix serves as the implementation of the final
resolved color-mix(), and is used for eager mixing (in the case
of the Raw variant currently used on workers) and in StyleColor
when doing the final resolution.

* Source/WebCore/css/color/CSSUnresolvedColor.cpp: Added.
(WebCore::CSSUnresolvedColor::cssText const):
(WebCore::CSSUnresolvedColor::equals const):
(WebCore::CSSUnresolvedColor::createStyleColor const):
* Source/WebCore/css/color/CSSUnresolvedColor.h: Added.
CSSUnresolvedColor serves a container for unresolved colors
in CSSPrimitiveValue. For now it only supports color-mix(),
but is setup to support other kinds going future.

* Source/WebCore/css/color/CSSUnresolvedColorMix.cpp: Added.
(WebCore::serializationForCSS):
(WebCore::operator==):
(WebCore::operator!=):
(WebCore::createStyleColor):
* Source/WebCore/css/color/CSSUnresolvedColorMix.h: Added.
Value type for unresolved color-mix(), used as the "specified
value" for color-mix() to correctly serialize per-spec and
correctly resolve to a structured StyleColor at style building
time.

* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeOriginColorRaw):
(WebCore::CSSPropertyParserHelpers::consumeOptionalAlphaRaw):
(WebCore::CSSPropertyParserHelpers::consumeOptionalAlphaRawAllowingSymbolTableIdent):
(WebCore::CSSPropertyParserHelpers::parseRelativeRGBParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseNonRelativeRGBParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseRGBParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseRelativeHSLParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseNonRelativeHSLParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseHSLParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseHWBParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseRelativeHWBParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseNonRelativeHWBParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseLabParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseRelativeLabParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseNonRelativeLabParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseLCHParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseRelativeLCHParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseNonRelativeLCHParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseColorFunctionForRGBTypesRaw):
(WebCore::CSSPropertyParserHelpers::parseRelativeColorFunctionForRGBTypes):
(WebCore::CSSPropertyParserHelpers::parseColorFunctionForXYZTypesRaw):
(WebCore::CSSPropertyParserHelpers::parseRelativeColorFunctionForXYZTypes):
(WebCore::CSSPropertyParserHelpers::parseRelativeColorFunctionParameters):
(WebCore::CSSPropertyParserHelpers::parseNonRelativeColorFunctionParameters):
(WebCore::CSSPropertyParserHelpers::parseColorFunctionParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseColorContrastFunctionParametersRaw):
(WebCore::CSSPropertyParserHelpers::consumeColorMixComponentRaw):
(WebCore::CSSPropertyParserHelpers::consumeColorMixComponent):
(WebCore::CSSPropertyParserHelpers::parseColorMixFunctionParametersRaw):
(WebCore::CSSPropertyParserHelpers::hasNonCalculatedZeroPercentage):
(WebCore::CSSPropertyParserHelpers::parseColorMixFunctionParameters):
(WebCore::CSSPropertyParserHelpers::parseColorFunctionRaw):
(WebCore::CSSPropertyParserHelpers::parseColorFunction):
(WebCore::CSSPropertyParserHelpers::consumeColorWorkerSafe):
(WebCore::CSSPropertyParserHelpers::consumeColor):
(WebCore::CSSPropertyParserHelpers::consumeOriginColor): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeOptionalAlpha): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeOptionalAlphaAllowingSymbolTableIdent): Deleted.
(WebCore::CSSPropertyParserHelpers::parseRelativeRGBParameters): Deleted.
(WebCore::CSSPropertyParserHelpers::parseNonRelativeRGBParameters): Deleted.
(WebCore::CSSPropertyParserHelpers::parseRGBParameters): Deleted.
(WebCore::CSSPropertyParserHelpers::parseRelativeHSLParameters): Deleted.
(WebCore::CSSPropertyParserHelpers::parseNonRelativeHSLParameters): Deleted.
(WebCore::CSSPropertyParserHelpers::parseHSLParameters): Deleted.
(WebCore::CSSPropertyParserHelpers::parseHWBParameters): Deleted.
(WebCore::CSSPropertyParserHelpers::parseRelativeHWBParameters): Deleted.
(WebCore::CSSPropertyParserHelpers::parseNonRelativeHWBParameters): Deleted.
(WebCore::CSSPropertyParserHelpers::parseLabParameters): Deleted.
(WebCore::CSSPropertyParserHelpers::parseRelativeLabParameters): Deleted.
(WebCore::CSSPropertyParserHelpers::parseNonRelativeLabParameters): Deleted.
(WebCore::CSSPropertyParserHelpers::parseLCHParameters): Deleted.
(WebCore::CSSPropertyParserHelpers::parseRelativeLCHParameters): Deleted.
(WebCore::CSSPropertyParserHelpers::parseNonRelativeLCHParameters): Deleted.
(WebCore::CSSPropertyParserHelpers::parseColorFunctionForRGBTypes): Deleted.
(WebCore::CSSPropertyParserHelpers::parseColorFunctionForXYZTypes): Deleted.
(WebCore::CSSPropertyParserHelpers::parseColorFunctionParameters): Deleted.
(WebCore::CSSPropertyParserHelpers::parseColorContrastFunctionParameters): Deleted.
(WebCore::CSSPropertyParserHelpers::normalizedMixPercentages): Deleted.
(WebCore::CSSPropertyParserHelpers::mixColorComponentsUsingColorInterpolationMethod): Deleted.
(WebCore::CSSPropertyParserHelpers::mixColorComponents): Deleted.
Renames color parsing functions to use the Raw suffix to match convention and
make it clear that the new color-mix parsing function is sufficiently different
in that it is not eagerly resolving.

* Source/WebCore/platform/graphics/ColorInterpolationMethod.cpp:
(WebCore::serializationForCSS):
* Source/WebCore/platform/graphics/ColorInterpolationMethod.h:
Add CSS serialization support and fix HasHueInterpolationMethod to actually
match the right structs.

* Source/WebCore/style/ColorFromPrimitiveValue.cpp:
(WebCore::Style::colorFromPrimitiveValue):
Support CSSUnresolvedColor CSSPrimitiveValues when building StyleColors.

Canonical link: https://commits.webkit.org/258970@main
